scipy/optimize/optimize.py

```diff
@@ -3061,6 +2912,576 @@ def show_options(solver=None, method=None, disp=True):
+        """
+        Advance the solver by a number of steps.
+
+        Parameters
+        ----------
+        iterations: int
+            The number of iterations to perform
+
+        Returns
+        -------
+        result: OptimizeResult
+            The optimization result represented as a OptimizeResult object.
+        """
+        # TODO interupt once convergence is reached?
+        for it in range(iterations):
+            self.x, self.fun = next(self)
+
+        self._finish_up()
+        return self.result
+
+    @property
+    def N(self):
+        """
+        The dimensionality of the problem, `np.size(x)`.
+        """
+        if self.x is not None:
+            return self.x.size
+        elif self.x0 is not None:
+            return self.x0.size
+        return RuntimeError("Cannot determine problem size at this time,"
+                            " perform at least one iteration")
+
+    def solve(self):
+        """
+        Run the solver through to completion.
+
+        Returns
+        -------
+        result: OptimizeResult
+            The optimization result represented as a `OptimizeResult` object.
+            Important attributes are: `x` the solution array, `success` a
+            Boolean flag indicating if the optimizer exited successfully and
+            `message` which describes the cause of the termination. See
+            `OptimizeResult` for a description of other attributes.
+        """
+        for x, fun in self:
+            self.x, self.fun = x, fun
+
+            if (self.converged() or self.nfev >= self.maxfun or
+                    self.nit >= self.maxiter):
+                break
+
+        self._finish_up()
+        return self.result
+
+    def converged(self):
+        """
+        The truth of whether the solver has converged.
+        """
+        raise NotImplementedError
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        """
+        Advance the solver by a single iteration.
+        """
+        # Should be over-ridden by each class based solver.
+        # TODO, Stop if self.maxfun/self.maxiter reached?
+        raise NotImplementedError
+    # step is because PyTorch has a step method
+    next = step = __next__
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        self._finish_up()
+
+    def _finish_up(self):
+        """
+        Each class based solver should override this to define end-of-run
+        actions. This includes setting the `message` status, `warn_flag`,
+        `success`, etc. It gets called at the end of the `solve` and `__call__`
+        methods, and during `__exit__` of the context manager.
+        """
+        pass
+
+    def show_options(self):
+        # each optimizer should return their own options text.
+        return ""
+
+
+class Function(object):
+    def __init__(self, func=None, args=(), jac=None, hess=None):
+        self._func = func
+        self.args = args
```

Comment:
Do we still want `args`? I remember it being a pain point. I don't think we want to remove it for backwards compatibility etc.

Reply:
Could we have some keyword arguments too?

Reply:
I don't see why not.

Reply:
Done